src/services/SearchService.ts

```diff
@@ -1,7 +1,281 @@
+          'metadata'
+        );
+        set.add({ space, id: v });
+      }
+
+      return this.findSpaces(Array.from(set));
+    }
+
+    return [];
   };
+
+  private async findSpaces(spaces): Promise<string[]> {
+    const needed = new Set<string>();
+
+    const dates = this.getAskDates();
+    if (!dates.length) return [];
+
+    for (const i of spaces) {
+      const space = i.space as Space;
+
+      //check rules
+      if (!(await this.checkRules(dates, i.id))) {
+        continue;
+      }
+
+      //check space capacity
+      if (!this.checkSuitableQuantity(space)) {
+        continue;
+      }
+
+      //check dates is available
+      if (await this.checkAvailableDates(i.id, dates)) {
+        needed.add(i.id);
+      }
+    }
+
+    return Array.from(needed);
+  }
+
+  private getAskDates(): DateTime[] {
+    if (!this.ask.checkIn || !this.ask.checkOut) {
+      throw ApiError.BadRequest('invalid dates in ask');
+    }
+
+    const checkIn = DateTime.fromFormat(this.checkInTime, 'hhmm', {
+      zone: this.facility.policies?.timezone
+    }).setZone('utc');
+
+    let from = DateTime.fromObject(
+      {
+        ...this.ask.checkIn,
+        hour: checkIn.hour,
+        minute: checkIn.minute
+      },
+      { zone: 'utc' }
+    );
+    const to = DateTime.fromObject(
+      {
+        ...this.ask.checkOut,
+        hour: checkIn.hour,
+        minute: checkIn.minute
+      },
+      { zone: 'utc' }
+    );
+
+    const dates: DateTime[] = [];
+
+    while (from <= to) {
+      dates.push(from);
+      from = from.plus({ days: 1 });
+    }
+
+    return dates;
+  }
+
+  private async checkRules(
+    dates: DateTime[],
+    spaceId: string
+  ): Promise<boolean> {
+    const spaceRuleRepository = new SpaceRuleRepository(
+      this.facilityId,
+      spaceId
+    );
+
+    const noticeRequirementRule = (await this.getRule(
+      'notice_required',
+      spaceRuleRepository
+    )) as NoticeRequiredRule;
+
+    const firstDay = dates[0];
+
+    const checkIn = DateTime.now().setZone('utc');
+    const interval = Interval.fromDateTimes(checkIn, firstDay).toDuration([
+      'days'
+    ]);
+
+    const intervalSeconds = convertDaysToSeconds(interval.get('days'));
+
+    if (
+      noticeRequirementRule &&
+      noticeRequirementRule.value > intervalSeconds
+    ) {
+      return false;
+    }
+
+    const formattedFirstDay = firstDay.toFormat('ccc').toLowerCase();
+    const lOSRule = (await this.getRule(
+      'length_of_stay',
+      spaceRuleRepository,
+      formattedFirstDay
+    )) as DayOfWeekLOSRule;
+
+    const lOS = dates.length;
+    let minLOS = 0;
+    let maxLOS = 0;
+
+    try {
+      minLOS = lOSRule[formattedFirstDay]?.minLengthOfStay || 0;
+    } catch (e) {
+      //rule not exist
+    }
+
+    try {
+      maxLOS = lOSRule[formattedFirstDay]?.maxLengthOfStay || 0;
+    } catch (e) {
+      //rule not exist
+    }
+
+    return (minLOS === 0 || lOS > minLOS) && (maxLOS === 0 || lOS < maxLOS);
```

Comment:
`minLOS === 0` and `maxLOS === 0` are redundant due to initialisation.

Reply:
fixed